### DISCUSSION TITLE
Incremental permission/scope requests for GitHub Apps

### DISCUSSION THREAD

**Discussion by birkjernstrom** [community/community#106748] [Apps, API and Webhooks]
### Select Topic Area

Product Feedback

### Body

First: Thanks for all the work you're all doing in building this amazing platform for developers 🙏🏼 It's great that you're building it as a true platform so that others can help extend it via Apps. I hope and believe this feedback can help accelerate such opportunities even further and greatly enrich the platform.

### Challenge
Currently, GitHub App developers can only request a global set of API permissions and directly upfront upon installation. Unfortunately, this creates significant challenges in building deep integrations and powerful extensions to GitHub without creating meaningful friction in the user experience. Since Apps offering more than one feature would require users to accept permissions required for all features upfront even though it's best that users can progressively opt-in to features.

### Example 

We're building polar.sh - a creator platform for developers & the open source ecosystem. We're proud to be building it [open source on GitHub](https://github.com/polarsource/polar) and as a deep integration to GitHub. Serving developers seeking to offer their community:
- Subscriptions of benefits native to our ecosystem, e.g Access to Private GitHub Repos (Requiring read/write to administration - inviting collaborators with appropriate permissions)
- Crowdfunding behind GitHub issues (Requiring read/write permissions to issues)
- Rewarding contributors from said funding (Requiring read permissions to PRs)
- (Idea for the future) Creating GitHub Discussions for their Posts/Newsletter written via Polar (Would require read/write permissions to Discussions)
- ... and a lot more in the pipeline

On Polar, all of these features are optional and something developers can progressively opt-in to use. It's up to them how they want to use Polar and what they want to offer their audience & communities.

However, unfortunately, we're forced to request **all** associated GitHub permissions needed to serve **all** of these optional features directly from developers. Resulting in 1) Unnecessary friction & concern for the user during onboarding & 2) Individual app installations (tokens) being used for 1 feature, but having permissions for all of them.

### Proposal

1. App developers still specifies **all** permissions required in their app settings for all their features (ceiling), but...
_Setting the maximum boundary conditions. Helpful to keep a cap & communicate all potential permissions required in the Marketplace etc._

2. Developers can gradually request to upgrade an installation with permissions within those boundaries, i.e incremental scope requests.
_Users would of course be required to grant such incremental requests on GitHub (as they should). Same flow as app installations currently._

I hope this is not too significant of a change. Since it seems like GitHub already stores individual scopes per installation given that existing installations are prompted to accept new permissions that the App adds to their global settings, but can deny and continue to use the installation with prior scopes/permissions.

### Alternative solutions are bad for everyone

1. **Requiring all permissions upfront & at once**
Covered extensively in the example and is the most common across GitHub apps today from what I can tell, unfortunately.

2. **Leverage multiple apps, i.e one GitHub App per feature/area**
Of course, it's possible to setup and use multiple Apps for each area of features offered. But it's bad for everyone:
   - User: Poor & confusing user experience to have N installations for the same service
   - GitHub: Multiple apps for the same service within your marketplace & service
   - Developer: Offering & maintaining multiple apps - brittle integrations & communication. Having N active access tokens per user etc.

In our case, today it would mean 2-3 apps. Given our backlog and ideas for the future that number would grow linearly for each new bigger feature set we want to support. It would be a terrible solution for all parties involved.

Looking forward to hearing from you and hopefully this is something we can work on together to greatly enhance the possibilities to build amazing Apps on GitHub.

**Comment by frankie567** [comment]
To add more background about this, it's a pattern that's supported and encouraged by Microsoft and Google (among others):

* https://learn.microsoft.com/en-us/security/zero-trust/develop/developer-strategy-authorization-best-practices#when-to-request-permission
* https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow#incrementalAuth

From what I understand, it was even supported by the legacy GitHub OAuth apps: https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps

---

Admittedly, GitHub Apps flow is not an OAuth2 flow, but we can probably have the same kind of behavior with a `permissions` query parameter that we could append to the installation URL.

**Comment by lior-orca** [comment]
Any updates on this one?
